File: part1/QuickUnion/QuickUnion.java
import java.util.Arrays;

public class QuickUnion {

    private int[] id;

    public QuickUnion(int N) {
        id = new int[N];
        for (int i = 0; i < N; i++) {
            id[i] = i;
        }
    }

    // 回溯到最上层的父节点
    private int root(int i) {
        while (i != id[i]) {
            i = id[i];
        }
        return i;
    }

    // 判断是否有同样的最上层父节点
    public boolean connected(int p, int q) {
        return root(p) == root(q);
    }

    public void union(int p, int q) {
        int i = root(p);
        int j = root(q);
        id[i] = j;
    }

    public static void main(String[] args) {
        QuickUnion qf = new QuickUnion(10);
        System.out.println(Arrays.toString(qf.id));
        qf.union(4, 3);
        System.out.println(Arrays.toString(qf.id));
        qf.union(3, 5);
        System.out.println(Arrays.toString(qf.id));
        qf.union(5, 6);
        System.out.println(Arrays.toString(qf.id));
        System.out.println(qf.connected(1, 9));
        System.out.println(qf.connected(5, 6));
    }
}

File: part1/QuickUnion/QuickUnion.js
class QuickUnion {
  id = [];

  constructor(N) {
    this.id = new Array(N);
    for (let i = 0; i < N; i++) {
      this.id[i] = i;
    }
  }

  root(i) {
    while (i !== this.id[i]) {
      i = this.id[i];
    }
    return i;
  }

  connected(p, q) {
    return this.root(p) === this.root(q);
  }

  union(p, q) {
    const rootp = this.root(p);
    const rootq = this.root(q);
    this.id[rootp] = rootq;
  }
}

const qu = new QuickUnion(10);
qu.union(4, 3);
qu.union(3, 8);
qu.union(6, 5);
qu.union(9, 4);
qu.union(2, 1);
console.log(qu.connected(8, 9));
console.log(qu.connected(5, 4));
qu.union(5, 0);
qu.union(7, 2);
qu.union(6, 1);
qu.union(7, 3);
console.log(qu.id);
console.log(qu.connected(8, 9));
